hydro_lang/src/sim/graph.rs

```diff
@@ -0,0 +1,578 @@
+};
+use crate::deploy::trybuild_rewriters::UseTestModeStaged;
+use crate::location::dynamic::LocationId;
+
+#[derive(Clone)]
+pub struct SimNode {}
+
+impl Node for SimNode {
+    type Port = ();
+    type Meta = ();
+    type InstantiateEnv = ();
+
+    fn next_port(&self) -> Self::Port {
+        todo!()
+    }
+
+    fn update_meta(&mut self, _meta: &Self::Meta) {}
+
+    fn instantiate(
+        &self,
+        _env: &mut Self::InstantiateEnv,
+        _meta: &mut Self::Meta,
+        _graph: DfirGraph,
+        _extra_stmts: Vec<syn::Stmt>,
+    ) {
+    }
+}
+
+#[derive(Clone)]
+pub struct SimExternal {
+    pub(crate) external_ports: Rc<RefCell<(Vec<usize>, usize)>>,
+    pub(crate) registered: RefCell<HashMap<usize, usize>>,
+}
+
+impl Node for SimExternal {
+    type Port = ();
+    type Meta = ();
+    type InstantiateEnv = ();
+
+    fn next_port(&self) -> Self::Port {
+        todo!()
+    }
+
+    fn update_meta(&mut self, _meta: &Self::Meta) {
+        todo!()
+    }
+
+    fn instantiate(
+        &self,
+        _env: &mut Self::InstantiateEnv,
+        _meta: &mut Self::Meta,
+        _graph: DfirGraph,
+        _extra_stmts: Vec<syn::Stmt>,
+    ) {
+    }
+}
+
+impl<'a> RegisterPort<'a, SimDeploy> for SimExternal {
+    fn register(&self, key: usize, port: usize) {
+        self.registered.borrow_mut().insert(key, port);
+    }
+
+    fn raw_port(&self, _key: usize) -> () {
+        todo!()
+    }
+
+    #[expect(clippy::manual_async_fn, reason = "false positive, involves lifetimes")]
+    fn as_bytes_bidi(
+        &self,
+        _key: usize,
+    ) -> impl Future<
+        Output = DynSourceSink<
+            Result<bytes::BytesMut, std::io::Error>,
+            bytes::Bytes,
+            std::io::Error,
+        >,
+    > + 'a {
+        async { todo!() }
+    }
+
+    #[expect(clippy::manual_async_fn, reason = "false positive, involves lifetimes")]
+    fn as_bincode_bidi<InT, OutT>(
+        &self,
+        _key: usize,
+    ) -> impl Future<Output = DynSourceSink<OutT, InT, std::io::Error>> + 'a
+    where
+        InT: serde::Serialize + 'static,
+        OutT: serde::de::DeserializeOwned + 'static,
+    {
+        async { todo!() }
+    }
+
+    #[expect(clippy::manual_async_fn, reason = "false positive, involves lifetimes")]
+    fn as_bincode_sink<T>(
+        &self,
+        _key: usize,
+    ) -> impl Future<Output = std::pin::Pin<Box<dyn futures::Sink<T, Error = std::io::Error>>>> + 'a
+    where
+        T: serde::Serialize + 'static,
+    {
+        async { todo!() }
+    }
+
+    #[expect(clippy::manual_async_fn, reason = "false positive, involves lifetimes")]
+    fn as_bincode_source<T>(
+        &self,
+        _key: usize,
+    ) -> impl Future<Output = std::pin::Pin<Box<dyn futures::Stream<Item = T>>>> + 'a
+    where
+        T: serde::de::DeserializeOwned + 'static,
+    {
+        async { todo!() }
+    }
+}
+
+pub(super) struct SimDeploy {}
+impl<'a> Deploy<'a> for SimDeploy {
```

Comment:
This doesn't do anything until the simulator supports networking, is that right?